### PR TITLE
rework keyboard input

### PR DESCRIPTION
**Pull request by filedesless**
basic keypress change detection and render distance controlled with J/K

### CODE REVIEW

**Comment by filedesless** [comment]
blocked by roadrunner-craft/core#19

**Comment by filedesless** [comment]
fix #12

File: src/render/renderer/renderer.rs [hunk 1]
```diff
@@ -9,8 +11,8 @@ pub struct Renderer {
 }
 
 impl Renderer {
-    pub fn update(&mut self, world: &World) {
-        self.chunk_renderer.update(world);
+    pub fn update(&mut self, world: &World, player: Vector3, keyboard: &KeyboardHandler) {
```
Comment (afrigon) [inline src/render/renderer/renderer.rs:14]: nit but a vector is not a player, it's a position or a player_position

File: src/render/renderer/chunk_renderer.rs [hunk 2]
```diff
@@ -146,17 +149,30 @@ impl ChunkRenderer {
...
 
-    pub fn update(&mut self, world: &World) {
-        // remove unloaded chunks' meshes
-        self.meshes
-            .retain(|coords, _| world.chunks.contains_key(coords));
+    pub fn update(&mut self, world: &World, player: Vector3, keyboard: &KeyboardHandler) {
+        if keyboard.just_pressed(VirtualKeyCode::J) && self.render_distance > 3 {
+            self.render_distance -= 1;
+        }
+        if keyboard.just_pressed(VirtualKeyCode::K) {
```
Comment (afrigon) [inline src/render/renderer/chunk_renderer.rs:160]: Make sure it doesn't go above LOAD_DISTANCE

**Comment by afrigon** [comment]
> blocked by [roadrunner-craft/core#19](https://github.com/roadrunner-craft/core/pull/19)

was merged

**Comment by afrigon** [comment]
input manager proxy should stay to allow controller input abstraction

**Comment by filedesless** [comment]
blocked by roadrunner-craft/core#20

File: src/input/input_handler.rs [hunk 3]
```diff
@@ -21,11 +21,23 @@ impl InputHandler {
...
         self.cursor.get_delta()
     }
 
     pub fn clear_cursor_delta(&mut self) {
         self.cursor.clear_delta()
     }
+
+    pub fn clear_keyboard_state(&mut self) {
+        self.keyboard.clear_keyboard_state()
+    }
```
Comment (afrigon) [inline src/input/input_handler.rs]: combine this into a clear method

File: src/input/keyboard.rs [hunk 4]
```diff
@@ -4,6 +4,8 @@ use std::collections::HashSet;
 #[derive(Debug, Default)]
 pub struct KeyboardHandler {
     pressed: HashSet<VirtualKeyCode>,
+    pressed_last_frame: HashSet<VirtualKeyCode>,
+    released_last_frame: HashSet<VirtualKeyCode>,
```
Comment (afrigon) [inline src/input/keyboard.rs:8]: c'est pas vraiment pressed last frame, more like pressed since cleared.

File: src/input/keyboard.rs [hunk 5]
```diff
@@ -19,13 +21,40 @@ impl KeyboardHandler {
...
+                    if !self.is_pressed(keycode) {
+                        self.pressed_last_frame.insert(keycode);
+                    }
+                    self.pressed.insert(keycode);
+                }
+                ElementState::Released => {
+                    if self.is_pressed(keycode) {
+                        self.released_last_frame.insert(keycode);
+                    }
+                    self.pressed.remove(&(keycode));
```
Comment (afrigon) [inline src/input/keyboard.rs:34]: jpense ya des parens superflus here ?

File: src/input/keyboard.rs [hunk 6]
```diff
@@ -19,13 +21,40 @@ impl KeyboardHandler {
...
+        self.pressed_last_frame.contains(&keycode)
+    }
+
+    /// indicates a previously pressed key was just released
+    pub fn just_released(&self, keycode: VirtualKeyCode) -> bool {
+        self.released_last_frame.contains(&keycode)
+    }
+
+    /// to call at the end of each gameloop
+    pub fn clear_keyboard_state(&mut self) {
```
Comment (afrigon) [inline src/input/keyboard.rs:56]: call this one just clear as well

File: src/render/renderer/chunk_renderer.rs [hunk 7]
```diff
@@ -180,8 +195,10 @@ impl ChunkRenderer {
                     .all(|other| !ChunkGridCoordinate::are_neighbours(coords, other))
         });
 
+        println!("meshes.len() => {}", self.meshes.len());
```
Comment (afrigon) [inline src/render/renderer/chunk_renderer.rs]: maybe remove this before merging ?

File: src/input/input_handler.rs [hunk 8]
```diff
@@ -21,11 +21,20 @@ impl InputHandler {
...
+
     pub fn get_cursor_delta(&self) -> &CursorDelta {
         self.cursor.get_delta()
     }
 
-    pub fn clear_cursor_delta(&mut self) {
-        self.cursor.clear_delta()
+    pub fn clear(&mut self) {
+        self.cursor.clear_delta();
+        self.keyboard.clear_keyboard_state();
```
Comment (afrigon) [inline src/input/input_handler.rs:38]: should we call these cleared as well ?

**Pull request merged**